app/access/tests/test_api.py

```diff
@@ -108,3 +106,82 @@ def test_get_users_returns_users_ordered_by_id(self):
                 },
             ]
         }
+
+    def test_post_users_creates_new_user_in_db_and_returns_it(self):
+
+        payload = {
+            "username": "donny",
+            "first_name": "Theodore Donald",
+            "last_name": "Kerabatsos",
+            "email": "donny@bowling.com",
+            "provider_id": Provider.objects.last().id,
+        }
+
+        response = self.client.post("users", json=payload)
+
+        assert response.status_code == 201
+        assert response.data == payload
+
+    def test_post_users_returns_404_if_provider_id_does_not_exist(self):
+
+        non_existing_provider_id = Provider.objects.last().id + 1
+        payload = {
+            "username": "donny",
+            "first_name": "Theodore Donald",
+            "last_name": "Kerabatsos",
+            "email": "donny@bowling.com",
+            "provider_id": non_existing_provider_id,
+        }
+
+        response = self.client.post("users", json=payload)
+
+        assert response.status_code == 404
+
+    def test_post_users_returns_422_if_email_format_invalid(self):
+
+        invalid_email = "donny_at_bowling_dot_com"
+        payload = {
+            "username": "donny",
+            "first_name": "Theodore Donald",
+            "last_name": "Kerabatsos",
+            "email": invalid_email,
+            "provider_id": Provider.objects.last().id,
+        }
+
+        response = self.client.post("users", json=payload)
+
+        assert response.status_code == 422
+        assert response.data == {'detail': "['Enter a valid email address.']"}
```

Comment:
Is there more in the response error payload or is this it? In the spec I have currently sth like this 
```
{
    "code": 409,
    "description": "Username is already taken"
}
```
I can change that, no problem. The `code` information is now arguably redundant to the HTTP header code. Question is if it makes sense to have unique codes (`E1234`) that can possibly occur somewhere in the whole platform..? Then a `code` would make sense.

Reply:
Yes, this is it, only a `detail` field for an uncaught `ninja.errors.HttpError`. Ok, so I keep it as it is.

Unique internal error code: I have no experience with that but I could imagine that this would be useful for support. I'll add as a discussion point. Out of scope for this PR.